clients/filesystem-fuse/src/open_dal_filesystem.rs

```diff
@@ -0,0 +1,269 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+use crate::config::AppConfig;
+use crate::error::ErrorCode::OpenDalError;
+use crate::filesystem::{
+    FileReader, FileStat, FileSystemCapacity, FileSystemContext, FileWriter, PathFileSystem, Result,
+};
+use crate::gvfs_fuse::FileSystemSchema;
+use crate::opened_file::{OpenFileFlags, OpenedFile};
+use crate::utils::GvfsResult;
+use async_trait::async_trait;
+use bytes::Bytes;
+use fuse3::FileType::{Directory, RegularFile};
+use fuse3::{Errno, FileType, Timestamp};
+use log::{debug, error};
+use opendal::layers::LoggingLayer;
+use opendal::services::S3;
+use opendal::{Builder, EntryMode, ErrorKind, Metadata, Operator};
+use std::path::Path;
+use std::time::SystemTime;
+
+pub(crate) struct OpenDalFileSystem {
+    op: Operator,
+}
+
+impl OpenDalFileSystem {}
+
+impl OpenDalFileSystem {
+    fn new(op: Operator, _config: &AppConfig, _fs_context: &FileSystemContext) -> Self {
+        Self { op: op }
+    }
+
+    pub(crate) fn create_file_system(
+        schema: &FileSystemSchema,
+        config: &AppConfig,
+        fs_context: &FileSystemContext,
+    ) -> GvfsResult<Box<dyn PathFileSystem>> {
+        match schema {
+            FileSystemSchema::S3 => {
+                let builder = S3::from_map(config.extent_config.clone());
+
+                let op = Operator::new(builder);
+                if let Err(e) = op {
+                    error!("opendal create failed: {:?}", e);
+                    return Err(OpenDalError.to_error(format!("opendal create failed: {:?}", e)));
+                }
+                let op = op.unwrap().layer(LoggingLayer::default()).finish();
+                Ok(Box::new(OpenDalFileSystem::new(op, config, fs_context)))
+            }
+        }
+    }
+
+    fn opendal_meta_to_file_stat(&self, meta: &Metadata, file_stat: &mut FileStat) {
+        let now = SystemTime::now();
+        let mtime = meta.last_modified().map(|x| x.into()).unwrap_or(now);
+
+        file_stat.size = meta.content_length();
+        file_stat.kind = opendal_filemode_to_filetype(meta.mode());
+        file_stat.ctime = Timestamp::from(mtime);
+        file_stat.atime = Timestamp::from(now);
+        file_stat.mtime = Timestamp::from(mtime);
+    }
+}
+
+#[async_trait]
+impl PathFileSystem for OpenDalFileSystem {
+    async fn init(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn stat(&self, path: &Path) -> Result<FileStat> {
+        let file_name = path.to_string_lossy().to_string();
+        let meta = self
+            .op
+            .stat_with(&file_name)
+            .await
+            .map_err(opendal_error_to_errno)?;
+
+        let mut file_stat = FileStat::new_file_filestat_with_path(path, 0);
+        self.opendal_meta_to_file_stat(&meta, &mut file_stat);
+        Ok(file_stat)
+    }
+
+    async fn read_dir(&self, path: &Path) -> Result<Vec<FileStat>> {
+        let file_name = path.to_string_lossy().to_string();
```

Comment:
`path` might be something other than a directory?

Reply:
That‘s impossible.  
Actually, It is impossible to distinguish a file from a directory based on the `path` alone. If the wrong type is passed, an error will occur later.